File: articles/calculus-for-systems-modeling/python/calculus_dynamic_simulation.py
"""
Calculus for Systems Modeling:
Numerical calculus and dynamic simulation.

Educational example only.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def simulate_logistic(
    initial_state: float,
    rate: float,
    capacity: float,
    dt: float,
    steps: int
) -> pd.DataFrame:
    """Simulate dS/dt = rS(1 - S/K) using Euler's method."""
    time = np.zeros(steps)
    state = np.zeros(steps)
    derivative = np.zeros(steps)

    state[0] = initial_state

    for i in range(1, steps):
        derivative[i - 1] = rate * state[i - 1] * (1.0 - state[i - 1] / capacity)
        state[i] = state[i - 1] + derivative[i - 1] * dt
        time[i] = time[i - 1] + dt

    derivative[-1] = rate * state[-1] * (1.0 - state[-1] / capacity)

    return pd.DataFrame({
        "time": time,
        "state": state,
        "derivative": derivative,
        "rate": rate,
        "capacity": capacity
    })


def cumulative_trapezoid(x: np.ndarray, y: np.ndarray) -> np.ndarray:
    """Approximate cumulative integral using the trapezoid rule."""
    cumulative = np.zeros_like(y)

    for i in range(1, len(y)):
        width = x[i] - x[i - 1]
        area = 0.5 * (y[i] + y[i - 1]) * width
        cumulative[i] = cumulative[i - 1] + area

    return cumulative


def main() -> None:
    output_dir = "../outputs"

    simulation = simulate_logistic(
        initial_state=10.0,
        rate=0.20,
        capacity=100.0,
        dt=0.1,
        steps=300
    )

    simulation["cumulative_state"] = cumulative_trapezoid(
        simulation["time"].to_numpy(),
        simulation["state"].to_numpy()
    )

    parameter_grid = pd.read_csv("../data/calculus_parameter_grid.csv")

    rows = []

    for _, row in parameter_grid.iterrows():
        result = simulate_logistic(
            initial_state=float(row["initial_state"]),
            rate=float(row["rate"]),
            capacity=float(row["capacity"]),
            dt=float(row["dt"]),
            steps=int(row["steps"])
        )

        rows.append({
            "rate": row["rate"],
            "capacity": row["capacity"],
            "final_state": float(result["state"].iloc[-1]),
            "maximum_state": float(result["state"].max()),
            "maximum_derivative": float(result["derivative"].max())
        })

    sensitivity = pd.DataFrame(rows).sort_values("final_state", ascending=False)

    print(simulation.head())
    print(sensitivity)

    simulation.to_csv(f"{output_dir}/calculus_dynamic_simulation.csv", index=False)
    sensitivity.to_csv(f"{output_dir}/calculus_sensitivity_summary.csv", index=False)


if __name__ == "__main__":
    main()

File: articles/calculus-for-systems-modeling/r/calculus_dynamic_simulation.R
# Calculus for Systems Modeling:
# Simulating continuous change in R.
# Educational example only.

library(tidyverse)

simulate_logistic <- function(initial_state, rate, capacity, dt, steps) {
  state <- numeric(steps)
  time <- numeric(steps)
  derivative <- numeric(steps)

  state[1] <- initial_state
  time[1] <- 0

  for (i in 2:steps) {
    derivative[i - 1] <- rate * state[i - 1] * (1 - state[i - 1] / capacity)
    state[i] <- state[i - 1] + derivative[i - 1] * dt
    time[i] <- time[i - 1] + dt
  }

  derivative[steps] <- rate * state[steps] * (1 - state[steps] / capacity)

  tibble(
    time = time,
    state = state,
    derivative = derivative,
    rate = rate,
    capacity = capacity
  )
}

parameter_grid <- read_csv("../data/calculus_parameter_grid.csv", show_col_types = FALSE)

simulation_results <- parameter_grid |>
  mutate(
    simulation = pmap(
      list(initial_state, rate, capacity, dt, steps),
      simulate_logistic
    )
  ) |>
  unnest(simulation)

summary_results <- simulation_results |>
  group_by(rate, capacity) |>
  summarise(
    final_state = state[time == max(time)],
    maximum_state = max(state),
    maximum_derivative = max(derivative),
    .groups = "drop"
  ) |>
  arrange(desc(final_state))

dir.create("../outputs", showWarnings = FALSE, recursive = TRUE)

write_csv(simulation_results, "../outputs/r_calculus_dynamic_simulation.csv")
write_csv(summary_results, "../outputs/r_calculus_sensitivity_summary.csv")

print(summary_results)
